### PR TITLE
sev-shim: Implement `allocate_and_map_memory()` and allocate the shim stack

### PR DESCRIPTION
**Pull request by haraldh**
Allocate memory and map it to a given virtual address for the shim stack.

### CODE REVIEW

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+pub struct ShimFrameAllocator {
+    max: usize,
+    next_4k: Address<usize, Page4KiB>,
+    next_2M: Address<usize, Page2MiB>,
+}
+
+impl core::fmt::Debug for ShimFrameAllocator {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+        f.debug_struct("ShimFrameAllocator")
+            .field("max   ", &format_args!("{:#?}", self.max as *const u8))
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:42]: yeah.. it was aligned with the other name lenghts :)

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+};
+use x86_64::{align_down, PhysAddr, VirtAddr};
+
+/// An aligned 2MiB Page
+///
+/// The `x86_64::structures::paging::Page<S>` is not aligned, so we use
+/// memory::Page as Page4KiB and this Page2MiB
+#[derive(Copy, Clone)]
+#[repr(C, align(2097152))]
+pub struct Page2MiB([u8; 2097152]);
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:19]: :notes:  "if it makes you happy..." :notes:

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    pub static ref SHIM_FRAME_ALLOCATOR_RWLOCK: RwLock<ShimFrameAllocator> = {
+        RwLock::<ShimFrameAllocator>::const_new(
+            spinning::RawRwLock::const_new(),
+            unsafe { ShimFrameAllocator::new() },
+        )
+    };
+}
+
+/// A frame allocator
+#[allow(non_snake_case)]
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:32]: :notes:  "if it makes you happy..." :notes:

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+pub struct ShimFrameAllocator {
+    max: usize,
+    next_4k: Address<usize, Page4KiB>,
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:35]: :notes:  "if it makes you happy..." :notes:

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    next_2M: Address<usize, Page2MiB>,
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:36]: :notes:  "if it makes you happy..." :notes:

File: enarx-keep-sev-shim/src/shim_stack.rs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+use x86_64::VirtAddr;
+
+/// The virtual address of the main kernel stack
+pub const SHIM_STACK_START: u64 = 0xFFFF_FF48_4800_0000;
+
+/// The size of the main kernel stack
+#[allow(clippy::integer_arithmetic)]
+pub const SHIM_STACK_SIZE: u64 = bytes![4; MiB];
+
+/// Allocat the stack for the shim with guard pages
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/shim_stack.rs:20]: :+1:

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+///
+/// The `x86_64::structures::paging::Page<S>` is not aligned, so we use
+/// memory::Page as Page4KiB and this Page2MiB
+#[derive(Copy, Clone)]
+#[repr(C, align(2097152))]
+pub struct Page2MiB([u8; 2097152]);
+
+lazy_static! {
+    /// The global ShimFrameAllocator RwLock
+    pub static ref SHIM_FRAME_ALLOCATOR_RWLOCK: RwLock<ShimFrameAllocator> = {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:23]: yes

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        RwLock::<ShimFrameAllocator>::const_new(
+            spinning::RawRwLock::const_new(),
+            unsafe { ShimFrameAllocator::new() },
+        )
+    };
+}
+
+/// A frame allocator
+#[allow(non_snake_case)]
+pub struct ShimFrameAllocator {
```
[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:33]: ah well...
```diff
-unsafe impl FrameAllocator<Size2MiB> for ShimFrameAllocator {
+unsafe impl paging::FrameAllocator<Size2MiB> for FrameAllocator {
```

traits and structs with the same name... at least nothing else needs to use FrameAllocator directly by name.

[the remark this answers is not in the document]
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs:33]: :notes: "if it makes you happy..." :notes:

File: enarx-keep-sev-shim/src/paging.rs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+/// Initialize a new OffsetPageTable.
+///
+/// This function is unsafe because the caller must guarantee that the
+/// complete physical memory is mapped to virtual memory at the passed
+/// `physical_memory_offset`. Also, this function must be only called once
+/// to avoid aliasing `&mut` references (which is undefined behavior).
+///
+/// # Safety
+///
+/// This function is unsafe because the caller must guarantee that the
```
Comment (ambaxter) [inline enarx-keep-sev-shim/src/paging.rs:29]: Is this duplicate text on purpose?

File: enarx-keep-sev-shim/src/shim_stack.rs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    dbg!(FRAME_ALLOCATOR.read().deref());
+
+    let mem_slice = FRAME_ALLOCATOR
+        .write()
+        .allocate_and_map_memory(
+            SHIM_PAGETABLE.write().deref_mut(),
+            VirtAddr::new(SHIM_STACK_START),
+            SHIM_STACK_SIZE as _,
+            PageTableFlags::PRESENT | PageTableFlags::WRITABLE | PageTableFlags::NO_EXECUTE,
+            PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
```
Comment (ambaxter) [inline enarx-keep-sev-shim/src/shim_stack.rs:46]: Are PRESENT and WRITABLE supposed to be doubled?

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 11]
```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! The global FrameAllocator
+use crate::BOOT_INFO;
+use memory::Page as Page4KiB;
```
Comment (ambaxter) [inline enarx-keep-sev-shim/src/frame_allocator.rs:5]: I think it makes sense to move Page2MiB into memory and refactor Page into Page4KiB, though I imagine that would balloon the PR quite a bit.

Comment (ambaxter) [reply enarx-keep-sev-shim/src/frame_allocator.rs:5]: @connorkuehl thoughts?

Comment (connorkuehl) [reply enarx-keep-sev-shim/src/frame_allocator.rs:5]: I'm not opposed to it, but I'm also happy to leave it as is as an implementation detail that the VM/SEV shim worries about; at least until the VM/SEV shim is not the only component that needs such functionality.

File: enarx-keep-sev-shim/src/paging.rs (hunk 9 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/paging.rs:29]: err... maybe not

File: enarx-keep-sev-shim/src/shim_stack.rs (hunk 10 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/shim_stack.rs:46]: doubled?

Comment (ambaxter) [reply enarx-keep-sev-shim/src/shim_stack.rs:46]: Nvm. I misread the code somewhat

File: enarx-keep-sev-shim/src/paging.rs (hunk 9 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/paging.rs:24]: corrected

**Pull request merged**